framework/model/module/module.go

```diff
@@ -24,67 +25,144 @@ type InitCallbacks struct {
 	AddStartup func(func(ctx context.Context))
 }
 
+type moduleConfig struct {
+	module Module
+	config configH
+}
+
+type configH struct {
+	config      *bootconfig.Config
+	generalJson []byte
+}
+
 type Module interface {
-	Name() string
+	Kind() string
 	Config() proto.Message
 	InitScheme(scheme *runtime.Scheme) error
 	InitManager(mgr manager.Manager, env bootstrap.Environment, cbs InitCallbacks) error
+	Clone() Module
 }
 
-func Main(bundle string, modules []Module) {
+func Main(modules []Module) {
+
 	fatal := func() {
 		os.Exit(1)
 	}
 
-	type configH struct {
-		config      *bootconfig.Config
-		generalJson []byte
+	// prepare module definition map
+	moduleDefinitions := make(map[string]Module)
+	for _, mod := range modules {
+		if _, ok := moduleDefinitions[mod.Kind()]; !ok {
+			moduleDefinitions[mod.Kind()] = mod
+			continue
+		}
 	}
 
+	// Init module of instance
+	var mcs []*moduleConfig
+
+	// get main module config
 	config, rawCfg, generalJson, err := bootstrap.GetModuleConfig("")
 	if err != nil {
 		panic(err)
 	}
 	if config == nil {
-		panic(fmt.Errorf("module config nil for %s", bundle))
+		panic(fmt.Errorf("main module config is nil"))
 	}
 	err = util.InitLog(config.Global.Log)
 	if err != nil {
 		panic(err)
 	}
+	log.Infof("load module config of %s: %s", config.Name, string(rawCfg))
 
-	log.Infof("load module config of %s: %s", bundle, string(rawCfg))
-	modConfigs := map[string]configH{}
+	// check if main module is bundle or not
 	isBundle := config.Bundle != nil
-	if isBundle {
+	if !isBundle {
+		var m Module
+		if config.Kind != "" {
+			m = moduleDefinitions[config.Kind].Clone()
+		} else {
+			// compatible for old version without kind field
+			m = moduleDefinitions[config.Name].Clone()
+		}
+		if m == nil {
+			log.Errorf("wrong kind or name of module %s", config.Name)
+			fatal()
+		}
+		mc := &moduleConfig{
+			module: m,
+			config: configH{
+				config:      config,
+				generalJson: generalJson,
+			},
+		}
+		mcs = append(mcs, mc)
+	} else {
 		for _, mod := range config.Bundle.Modules {
+			//log.Warningf("mod: %+v", mod)
+			var m Module
+			if mod.Kind != "" {
+				m = moduleDefinitions[mod.Kind].Clone()
+			} else {
+				// compatible for old version without kind field
+				m = moduleDefinitions[mod.Name].Clone()
+			}
+			if m == nil {
+				log.Errorf("wrong kind or name of module %s", mod.Name)
+				fatal()
+			}
+
 			modConfig, modRawCfg, modGeneralJson, err := bootstrap.GetModuleConfig(mod.Name)
 			if err != nil {
 				panic(err)
 			}
-			if config == nil {
+			if modConfig == nil {
 				panic(fmt.Errorf("module config nil for %s", mod.Name))
```

Comment:
I wanna know if there're some modules which can work without config.

Reply:
Changed these code. submodule without config is allowed, because user can specify some global config in main module config.